### ISSUE TITLE
toml fails to compile due to IndexMap

### ISSUE THREAD

**Issue by andrewbaxter** [toml-rs/toml#539]
re #524

I think this issue also occurs on `toml` per: https://github.com/andrewbaxter/genemichaels/issues/72

I wasn't able to reproduce it unfortunately, `toml` is a transitive dep.  After patching in a fork of the parent dep though I saw both indexmap feature "default" and "std" appearing in the cargo tree separately:

```
│       ├── toml feature "default"
│       │   ├── toml v0.7.3
│       │   │   ├── serde feature "default" (*)
│       │   │   ├── indexmap feature "default"
│       │   │   │   └── indexmap v1.9.3
│       │   │   │       └── hashbrown feature "raw"
│       │   │   │           └── hashbrown v0.12.3
│       │   │   │       [build-dependencies]
│       │   │   │       └── autocfg feature "default"
│       │   │   │           └── autocfg v1.1.0
│       │   │   ├── serde_spanned feature "default"
│       │   │   │   └── serde_spanned v0.6.1
│       │   │   │       └── serde feature "default" (*)
│       │   │   ├── serde_spanned feature "serde"
│       │   │   │   └── serde_spanned v0.6.1 (*)
│       │   │   ├── toml_datetime feature "default"
│       │   │   │   └── toml_datetime v0.6.1
│       │   │   │       └── serde feature "default" (*)
│       │   │   ├── toml_datetime feature "serde"
│       │   │   │   └── toml_datetime v0.6.1 (*)
│       │   │   ├── toml_edit feature "default"
│       │   │   │   └── toml_edit v0.19.8
│       │   │   │       ├── serde feature "default" (*)
│       │   │   │       ├── indexmap feature "default" (*)
│       │   │   │       ├── indexmap feature "std"
│       │   │   │       │   └── indexmap v1.9.3 (*)
│       │   │   │       ├── serde_spanned feature "default" (*)
│       │   │   │       ├── serde_spanned feature "serde" (*)
│       │   │   │       ├── toml_datetime feature "default" (*)
│       │   │   │       └── winnow feature "default"
│       │   │   │           ├── winnow v0.4.1
│       │   │   │           └── winnow feature "std"
│       │   │   │               ├── winnow v0.4.1
│       │   │   │               └── winnow feature "alloc"
│       │   │   │                   └── winnow v0.4.1
│       │   │   └── toml_edit feature "serde"
│       │   │       ├── toml_edit v0.19.8 (*)
│       │   │       └── toml_datetime feature "serde" (*)
│       │   ├── toml feature "display"
│       │   │   └── toml v0.7.3 (*)
│       │   └── toml feature "parse"
│       │       └── toml v0.7.3 (*)
```

(Sorry if this is off base though, I'm not even sure how autocfg is included by indexmap - AFAICT it doesn't even appear in the Cargo.toml so I may be out of my depth)

**Comment by epage** [comment]
Is that `cargo tree` output for a run that failed or succeeded?

While `toml` should also set it, you should be getting it through `toml_edit`.  Note that all of this is dependent on the `Cargo.lock` so when helping someone through this, their `cargo tree` is important.  Patching dependencies should not be needed.  They can use `cargo update` to control specific dependencies, like `cargo update -p toml_edit --precise 0.19.8`.

**Comment by epage** [comment]
Note that for `toml`, this is only a problem if
- The `preserve_order ` feature is enabled
- (The `parse` and `display` features are disabled) or using an older `toml_edit`

**Comment by andrewbaxter** [comment]
That was a tree output from the successful build I ran locally.  Hmm, okay, maybe I was reading the tree output wrong - I thought there were two `indexmap` deps included, one with the feature and one without.  Please feel free to close since it seems like the issue isn't here, I do appreciate the help though.  Not sure it's useful, but here's the reverse tree from the indexmap dep:
```
$ cargo tree -i indexmap --edges features
indexmap v1.9.3
├── indexmap feature "default"
│   ├── toml v0.7.3
│   │   ├── toml feature "default"
│   │   │   └── cargo-manifest v0.7.1 (/home/andrew/temp/soft/fork-cargo-manifest)
│   │   │       └── cargo-manifest feature "default" (command-line)
│   │   ├── toml feature "display"
│   │   │   └── toml feature "default" (*)
│   │   ├── toml feature "indexmap"
│   │   │   └── toml feature "preserve_order"
│   │   │       └── cargo-manifest v0.7.1 (/home/andrew/temp/soft/fork-cargo-manifest) (*)
│   │   ├── toml feature "parse"
│   │   │   └── toml feature "default" (*)
│   │   └── toml feature "preserve_order" (*)
│   └── toml_edit v0.19.8
│       ├── toml_edit feature "default"
│       │   └── toml v0.7.3 (*)
│       └── toml_edit feature "serde"
│           └── toml v0.7.3 (*)
└── indexmap feature "std"
    └── toml_edit v0.19.8 (*)
```

The `cargo update -p toml_edit` thing only works if that's a direct dependency, right? `toml` is a dep of `cargo-manifest` which is the direct dependency of `genemichaels`, so I forked `cargo-manifest` to update the `toml` dep to `0.7.3`, which included the `toml_edit` package on it's own.  `cargo-manifest` (patched from 0.5.6) has:
```
toml = { version = "0.7.3", features = ["preserve_order"] }
```
So I think the `parse` and `display` features should be included (not disabling default features).

**Comment by andrewbaxter** [comment]
Okay, so to summarize (for my own sake)
1. The original build was with `toml-0.5.10`.  This pulled in `indexmap-1.9.3`.  `toml-0.5.10` doesn't touch the indexmap features... `indexmap-1.9.3` seems to have `std` on by default but maybe that's wrong.
2. If I update to `toml-0.7.3`, that includes `toml_edit` which should set the `std` flag on `indexmap`.  `toml` itself also includes `indexmap`, with only the default features, but I guess cargo merges deps even if their features are different.

So updating to `0.7.3` should be sufficient...
